Doc/library/typing.rst

```diff
@@ -572,6 +572,32 @@ These can be used as types in annotations and do not support ``[]``.
    * Every type is compatible with :data:`Any`.
    * :data:`Any` is compatible with every type.
 
+.. data:: Never
+
+   Notation for the `bottom type <https://en.wikipedia.org/wiki/Bottom_type>`_,
+   a type that has no members.
+
+   This can be used to define a function that should never be
+   called, or a function that never returns::
+
+     from typing import Never
+
+     def never_call_me(arg: Never) -> None:
+         pass
+
+     never_call_me(1)  # type checker error
+
+     def stop() -> Never:
+         return 1  # type checker error
+
+   The :func:`assert_never()` function uses the ``Never`` type to statically
+   assert that code is unreachable.
+
+   .. versionadded:: 3.11
+
+      On older Python versions, :data:`NoReturn` may be used to express the
+      same concept. ``Never`` was added to make the intended meaning more explicit.
+
 .. data:: NoReturn
```

Comment:
Should we deprecate it?

Reply:
@erictraut's suggestion was to allow type checkers to limit NoReturn to what it's meant for (functions that don't return) in the future.

Reply:
> limit NoReturn to what it's meant for (functions that don't return) in the future.

Introducing such a type checker limitation (which I am ±0 on) would make it impossible to spell "bottom type" in code that must support Python <3.11, unless we **also add typing_extensions.Never**, which I do recommend.

Reply:
I'm also not sure we should do that, but it's nice to have the option.

I will add `typing_extensions.Never` once this PR is merged. (With some luck it will be a very feature-packed typing-extensions release: PEP 646, `@dataclass_transform`, and all the little functions I'm adding.

Reply:
> I'm also not sure we should do that, but it's nice to have the option.

Do what? Deprecate using `NoReturn` outside return types?

Honestly I feel there's very little reason to deprecate that given that it apparently used to work. Too much churn isn't good.

Reply:
Yes, that would be the suggestion, and agree that it's not a very valuable change. Sounds like there's no action item here, so I'll resolve this thread.